File: 2024/day07_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


from typing import Any, Literal

type Operation = Literal["SUM"] | Literal["MUL"]

type Equation = tuple[int, list[int]]


def solution(content: list[str]) -> Any:
    return sum(get_result(x) for x in (parse(line) for line in content) if test(x))


def get_result(eq: Equation) -> int:
    result, _ = eq
    return result


def test(equation: Equation) -> bool:
    result, args = equation
    for ops in operations(len(args) - 1):
        if apply(args, ops) == result:
            return True
    return False


def apply(args: list[int], ops: list[Operation]) -> int:
    acc, *rest = args
    for val, op in zip(rest, ops):
        acc = calc(acc, val, op)
    return acc


def calc(x: int, y: int, op: Operation) -> int:
    if op == "MUL":
        return x * y
    if op == "SUM":
        return x + y


def operations(n: int):
    return [bin_to_op(n, bin(x)) for x in range(2**n)]


def bin_to_op(n: int, seq: str) -> list[Operation]:
    return ["SUM" if x == "0" else "MUL" for x in seq[2:].rjust(n, "0")]


def parse(line: str) -> Equation:
    fst, *rest = line.split(" ")
    return (int(fst[:-1]), [int(x) for x in rest])


def main() -> None:
    with open("day07.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2024/day07-1.hs
import Data.List (find)
import Data.Maybe (isJust)

data Operation = MUL | SUM

type Equation = (Int, [Int])

solution :: String -> Int
solution = sum . map getResult . filter test . map parse . lines

getResult :: Equation -> Int
getResult = fst

test :: Equation -> Bool
test (res, args) = isJust . find (\ops -> res == apply args ops) $ operations (length args - 1)

apply :: [Int] -> [Operation] -> Int
apply (x : xs) ops = foldl (\x (y, op) -> calc x y op) x $ zip xs ops

calc :: Int -> Int -> Operation -> Int
calc x y SUM = x + y
calc x y MUL = x * y

operations :: Int -> [[Operation]]
operations 0 = [[]]
operations n = map (SUM :) recursive ++ map (MUL :) recursive
  where
    recursive = operations (n - 1)

parse :: String -> Equation
parse line =
  ( read . init . head . words $ line,
    map read . tail . words $ line
  )

main :: IO ()
main = do readFile "day07.dat" >>= print . solution
